clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -515,312 +448,319 @@ private void needsCoordinator(FindCoordinatorRequest.CoordinatorType type, Strin
-    }
-
-    private TransactionalRequest txnOffsetCommitRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
-                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+    private TxnRequestHandler txnOffsetCommitHandler(Map<TopicPartition, OffsetAndMetadata> offsets, String consumerGroupId) {
         for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
             OffsetAndMetadata offsetAndMetadata = entry.getValue();
-            pendingTxnOffsetCommits.put(entry.getKey(),
-                    new TxnOffsetCommitRequest.CommittedOffset(offsetAndMetadata.offset(), offsetAndMetadata.metadata()));
+            CommittedOffset committedOffset = new CommittedOffset(offsetAndMetadata.offset(), offsetAndMetadata.metadata());
+            pendingTxnOffsetCommits.put(entry.getKey(), committedOffset);
         }
-        return txnOffsetCommitRequest(consumerGroupId, isRetry, result);
-    }
 
-    private TransactionalRequest txnOffsetCommitRequest(String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
         TxnOffsetCommitRequest.Builder builder = new TxnOffsetCommitRequest.Builder(consumerGroupId,
-                pidAndEpoch.producerId, pidAndEpoch.epoch, OffsetCommitRequest.DEFAULT_RETENTION_TIME, pendingTxnOffsetCommits);
-        return new TransactionalRequest(builder, new TxnOffsetCommitCallback(consumerGroupId, result),
-                FindCoordinatorRequest.CoordinatorType.GROUP, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, consumerGroupId, result);
+                pidAndEpoch.producerId, pidAndEpoch.epoch, OffsetCommitRequest.DEFAULT_RETENTION_TIME,
+                pendingTxnOffsetCommits);
+        return new TxnOffsetCommitHandler(builder);
     }
 
-    private abstract class TransactionalRequestCallBack implements RequestCompletionHandler {
-        protected final TransactionalRequestResult result;
+    abstract class TxnRequestHandler<S extends AbstractRequest.Builder<?>, T extends AbstractResponse>
+            implements RequestCompletionHandler {
+        protected final TransactionalRequestResult result = new TransactionalRequestResult();
+        protected final S builder;
+        private boolean isRetry = false;
 
-        TransactionalRequestCallBack(TransactionalRequestResult result) {
-            this.result = result;
+        private TxnRequestHandler(S builder) {
+            this.builder = builder;
+        }
+
+        AbstractRequest.Builder<?> requestBuilder() {
+            return builder;
+        }
+
+        void fatal(RuntimeException e) {
+            result.setError(e);
+            result.done();
+            transitionTo(State.ERROR, e);
+        }
+
+        void fenced() {
+            log.error("Epoch has become invalid: producerId: {}. epoch: {}. Message: {}", pidAndEpoch.producerId,
```

Comment:
Maybe `Producer epoch...`. Also, not sure the exception message adds anything given what's already logged. Maybe we should remove that.